---
title: "post_DADA2_filtering"
author: "Nathan R. Geraldi"
date: "March 21, 2019"
output: github_document
---

set table options
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## libraries
```{r libraries, message=FALSE, warning=FALSE}
library(RColorBrewer)
library(rdrop2)  # what is this used for?
library(vegan)
library(fields)
library(psych)
library(tidyverse) 
library(broom)
library(metacoder)
library(rphylopic)

#  install.packages("devtools")
#  devtools::install_github("grunwaldlab/metacoder")

```

## functions
```{r functions, message=FALSE, warning=FALSE}
# function to remove rows with n number of NA's
delete.na <- function(DF, n=0) {
  DF[rowSums(is.na(DF)) <= n,]
}

# Specify alpha as a percentage:
colorRampAlpha <- function(..., n, alpha) {
   colors <- colorRampPalette(...)(n)
   paste(colors, sprintf("%x", ceiling(255*alpha)), sep="")
}

##############################################################################
## function to calulate dendogram which is then used in stratiplots
 #       from http://paleoecologie.umontreal.ca/category/code/
#   x should be dataframe spread so species are in columns
# example.......
#euk_phylum_dist<- euk_phylum2 %>% 
#  select(ybp,habitat,core,primer,Phylum,percent_reads) %>% 
#  filter(ybp<850) %>% 
#  spread(key=Phylum, value=percent_reads, fill=0) 

# for test but first run relevent code
# euk_phylum_ddata<-dendo_prep(x=euk_phylum_dist)

dendo_prep <- function(x, max_age = max_age_depth, first_dat_col = 5) {

age<-dplyr::pull(x, age_dat_col)  # can use name too

diss <- dist(sqrt(x[first_dat_col:length(names(x))]))
clust <- rioja::chclust(diss, method="coniss")  #  use ng= to specify howmany groups to display
# bstick(euk_clust, 20)   # optionnally look at number of optimal zones -where curves level out
# plot(euk_clust, ng=5) # limit groups but doesn't work

#Now we will use ggdendro and modify its output in order to inject ages that will replace sample numbers:
dendro <- as.dendrogram(clust)
ddata <- ggdendro::dendro_data(dendro, type="rectangle")
 
## MOD the segment data frame for age instead of value order -----------------------
ddata$segments->yoo
yoo$xx=NA
yoo$xxend=NA
 
for(i in 1:length(age)){
 yoo$xx[which(yoo$x == i)]=age[i]
 yoo$xxend[which(yoo$xend == i)]=age[i]
  yoo_da=age[i+1]-age[i]
 yoo$xx[yoo$x>i & yoo$x<i+1]<-age[i]+(yoo$x[yoo$x>i & yoo$x<i+1]-i)* yoo_da
 yoo$xxend[yoo$xend>i & yoo$xend<i+1]<-age[i]+(yoo$xend[yoo$xend>i & yoo$xend<i+1]-i)* yoo_da
} 
# Please dont ask....
ddata$segments$x<-yoo$xx
ddata$segments$xend<-yoo$xxend

return(ddata) # last line sets what is returned
}

##############################################################################
## to change location and size of legend in ggplot  -- not used in this script

change_legend <- function(myPlot, pointSize = 2, textSize = 9, spaceLegend = 0.1) {
    myPlot +
        guides(#fill = guide_legend(override.aes = list(size = pointSize)),
                fill = guide_colourbar(barwidth = 0.9, barheight = 4.0)) +
         #      color = guide_legend(override.aes = list(size = pointSize))) +
        theme(legend.position = c(0.44, 0.5), # center of legend
              legend.title = element_text(size = textSize), 
              legend.text  = element_text(size = textSize),
              plot.margin = unit(c(0.0,0,0.0,0), "lines") ) # top, right, bottom, left)
             # legend.key.size = unit(spaceLegend, "lines"))
}

##############################################################################
## to both reverse and log transfom axis in ggplot, need scales package
reverse_log_trans <- function(base = exp(1)) {
    trans <- function(x) -log(x, base)
    inv <- function(x) base^(-x)
    scales::trans_new(paste0("reverselog-", format(base)), trans, inv, 
              scales::log_breaks(base = base), 
              domain = c(1e-100, Inf))
}

## same but for square root trans
reverse_sqrt_trans <- function() {
    scales::trans_new(
        name = "rev_sqrt", 
        transform = function(x) -sqrt(abs(x)), 
        inverse = function(x) x^2);
}


```


## plot_function

```{r plot_func}
#   from http://paleoecologie.umontreal.ca/category/code/
# first plots top 7 in stratiplot then plots dendogram cluster
# x is datafram of top 7 and y is from dendo_prep function
#  taxa_group is usually Phylum or Genus and will define each individual stratiplot (eg. name of x$Phylum)
strati_plot_7 <- function(x, y, taxa_group, pic_rank, max_age = max_age_depth, abund_lim = c(0, 110),
                          abund_breaks = c(1,10,40,90), x_breaks = c(1,25,100,250,500,750), 
                          age_axis_trans="yes", add_pic="yes",
                          cluster_adj= 0) {
# cluster_adj was added to adjust vertical space, but shouldn't need it
  
  ## example - for test
  #       x=euk_phylum_top7; y=euk_phylum_ddata; taxa_group=euk_phylum_top7$Phylum; pic_rank= euk_phylum_rank;  max_age = max_age_depth; abund_lim = c(0, 110); abund_breaks = c(1,10,40,90); age_axis_trans="yes" ; x_breaks = c(1,25,100,250,500,750); cluster_adj= 0
  
 x$taxa_group1<-taxa_group ## set lineage level either Phylum or Genus range(x$ybp)
 x$age_dat<-dplyr::pull(x, age_dat_col) # didn't work -x$age_dat<-x[,age_dat_col]
  
  # plot stratiplot of top 7 
sp7<-ggplot(x, aes(x = age_dat, y = percent_reads)) +
   geom_point(aes(x = age_dat, y = percent_reads, colour=habitat, shape=primer), size = 2) +
   geom_smooth(aes(x = age_dat, y = percent_reads), span = 0.9, colour="grey", se=TRUE)   +
   geom_smooth(aes(x = age_dat, y = percent_reads, colour=habitat), span = 0.9, se=FALSE)   + #loess  se=FALSE
   ## if less than 1000 observations uses loess
  # facet by taxon, keeping distance on the x axis comparable
  facet_grid(~taxa_group1, scales = "fixed", space = "fixed") +
  # have the same breaks for all x axes
  scale_y_continuous(limits = abund_lim,  expand = c(0,0), breaks=abund_breaks , trans="sqrt", minor_breaks=NULL) + {
  ##
  # use wither transformed on raw age data for x axis
   #+  # for non transfomrewd axis - on left after flip
 # for both transformed and reversed axis reverse_sqrt_trans() or reverse_log_trans(10)  
      if(age_axis_trans=="yes") scale_x_continuous(trans=reverse_sqrt_trans(), limits = c(max_age, 0), breaks = x_breaks, expand = c(0.03,0), minor_breaks=NULL) else scale_x_reverse(limits = c(max_age, 0) ) 
    } +
  ##
  scale_color_manual(values=c("tan3","seagreen3")) +
  #scale_colour_brewer(palette = "Set1") +
  scale_shape(solid=FALSE) +
    labs(y = "Relative abundance (%)", x = "Years before present")  +
  coord_flip() +
   theme(legend.position="right") +
    theme_bw() + # classic -no grids or bw
  # increase margins and reduce font of plot labels wihtin facet_grid
  theme(panel.spacing = unit(0.5, "lines"), strip.text.x = element_text(size = 7),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf) +   # draw x andy axis, x on left - annotate("segment", x=-Inf, xend=-Inf, y=-Inf, yend=Inf)
  annotate("segment", x=Inf, xend=Inf, y=-Inf, yend=Inf) 


## plot dendogram     # mes<-y$segments
dendro<-ggplot(ggdendro::segment(y)) +
 geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +  
 coord_flip() +  # match with above xlim=c(0,850),ylim=c(600,1000)
 ylab("Distance") +  {
  #  !! if need to adjust cluster heigth !!!, a (+) will raise last cluster , needs to be reverse order-0 second
  ## use only one of following lines depending on transform
        if(age_axis_trans=="yes")   scale_x_continuous(trans=reverse_sqrt_trans(), limits = c(max(y$segments$xend) + cluster_adj*max(y$segments$xend),0), expand = c(0.03,0) )  else scale_x_reverse(limits=c(max(y$segments$xend) + cluster_adj*max(y$segments$xend),0))
    } +

   #theme(plot.margin = unit(c(1,0,1,1), "lines" )) +
   #theme(panel.grid.major = element_blank())  + # remove grids
 theme( panel.background = element_blank(),
 axis.title.y=element_blank(),
 axis.text.y=element_blank(),
 axis.ticks.y=element_blank(),
 axis.line.y = element_blank() ) +
    #annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)    # x on left - 
  annotate("segment", x=Inf, xend=Inf, y=-Inf, yend=Inf)  

## ADD LEGEND TO SP7
legend <- lemon::g_legend(sp7 + theme(legend.position='right')) 

## conver to Grob
g1 = ggplotGrob(sp7+theme(legend.position='hidden'))
g2 = ggplotGrob(dendro)

## align height
g2$heights[5:7]<-g1$heights[6:8] # Important to align plots!

## put together all plots
finplot<-ggpubr::ggarrange(g1, g2, legend, widths =c(1,0.2,0.11), ncol=3)
#  ,ncol=2,newpage = TRUE

    
    
  # the following is to add pictures from pic , added in import data
   # based on phyloseq, sheet needs taxonomy and id or picture
    #   pic_x=900, pic_y=40, pic_size=100, pic_color ="red" "grey20"
      if(add_pic=="yes"){
        
        pic_dat<- pic %>% 
          filter(taxa %in% pic_rank) %>% 
          mutate(taxa = factor(taxa, levels = pic_rank)) %>%
          arrange(taxa) %>% 
          mutate(x=seq(from=0.09, to=0.72, length.out = 7) ) %>%  # get x pos
          mutate(y = -0.32) %>% 
          mutate(wid = 0.04 * size_per_adj)
        
        finplot_2<-finplot # make new
        
       for (i in c(1:length(pic_dat[,1]))){
        img <- image_data(pic_dat$id[i], size = "256")[[1]]  # 256 or 512  ? some don't have 512
        finplot_2 <- cowplot::ggdraw(finplot_2) + 
  cowplot::draw_image(img, x=pic_dat$x[i], y=pic_dat$y[i], width=pic_dat$wid[i], hjust=0, vjust=0,scale = 1)
       }
      finplot<-finplot_2  
      }

        
### draw pic    failures - anything that uses existing axis  ### - don't try again!!
  #sp7 + add_phylopic(img, x=50, y=50, ysize=300, color=pic_color) # x/y-centers -doesn't with transformed axis
  # a1 = annotation_custom(rasterGrob(img, interpolate=TRUE), xmin=500, xmax=150, ymin=4, ymax=60)
#sp7 + a1

finplot

}
```



## universal variables
```{r define_universal}

## set some other universal variables
####  !!!! need to be in alphabetical order - tabs in sam_file should match these names too !!!!! also have "type"" column
primers<-c('18s_stoeck','18smini',"euka02","co1", "rbclmini","vert")
n_primers<-length(primers)

####  !!!! set minimum reads per sample to remove beforre mds   !!!!!   ####
min_reads_per_miseq<- 1000

################### set file path and name of sample data
## each primer should have own sheet with name matching primers
## primer sheets first then location sheet (1 row per each sample location), then other sheets with relevent data
sam_file_path<-"/Users/geraldn/Dropbox/Documents/KAUST/eDNA/Samples_Data/Dammam/Dammam_sample_data_all.xlsx" ## set sample data file
##


#####################   limit data by taxa  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
#  set "yes" if you want to remove specific taxa, yes will remove them
# this runs chunk-remove edaphic and different anmes for export
edaphic<-"yes"  #  edaphic<-"yes" 
#  list taxa to remove, keep separate for each taxa column to ensure not conficts
#            (instead of single vector using lineage)
kingdom_remove<-c("Fungi")
phylum_remove<-c("Annelida","Nematoda","Platyhelminthes","Rotifera","Nemertea")
order_remove<-c("Lucinoida","Nuculoida","Lamiales","Alismatales")

#####################      for plots   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
###    **********************************************
## TO DO specify column to plot on y axis -  e.g. ybp or core depth

## set max age or depth of plots
##       if adding pics will need to add more space at bottom
max_age_depth<-925   #  
max_age_depth_char<-as.character(max_age_depth)  # for plot names
age_dat_col<-1   #  column number (1 or 2 likely) of age or depth you want plotted  e.g. ypb or depth - check plot_euk
age_dat_col<-"ybp"  # use name it is safer !!!!!

 #  TO DO  either do by name, this needs to be worked on through dats tidy - but only matters at plot- call of function - strati_plot_7
```

## set paths and filenames
```{r paths}
stud_pat<-"Dammam"  # matches study specific title from pipe (begining of files).
dir<-"/Users/geraldn/Dropbox/"
out_file<-"Documents/KAUST/eDNA/R/pipe_summary"
# export  to project folder
export_file<-"Documents/KAUST/eDNA/R/csv/"

#  name of data
dat_name<-paste(stud_pat,"_filtered_data_all.csv",sep="")  
# name for dating info
dating_name<-paste(stud_pat,"_predicted_dating.csv",sep="")


# plot export file
if ( edaphic== "no") {
plot_file<-"Documents/KAUST/eDNA/R/plots/Dammam_core/strati/"
} else if (edaphic== "yes") {
plot_file<-"Documents/KAUST/eDNA/R/plots/Dammam_core/strati/no_surface_edaphic"
}

# location of excel with phylo pics
pic_file<-paste(dir,"Documents/R docs/phylopic/eDNA_phylo_pics.xlsx", sep="")

```


## import data

```{r import}
# sample data -- will need Quality control !!! make sure sample_se make sense
sheets <- openxlsx::getSheetNames(sam_file_path)
sam_dat <- lapply(sheets,openxlsx::read.xlsx,xlsxFile=sam_file_path)  # mes1<-sam_dat[[2]]   names(mes1)
names(sam_dat) <- sheets   # add name to each list
##    move to next .Rmd - nothing to do her
locations<-sam_dat[[n_primers+1]]  # isolate locations
sample_dat<-sam_dat[[n_primers+2]]  # isolate other data
sample_age<-sam_dat[[n_primers+3]]  # isolate other age
##
sam_dat <-sam_dat[1:n_primers]   # keep only primer smaple data
sam_dat <- lapply(sam_dat, function(df) mutate_at(df, .vars="sample_ID", as.character))  # make sure sample_ID is character
sample_sam<-bind_rows(sam_dat, .id = 'id')  # names(sam_dat[1:6]) 

# predicted dates
dates_pred<-data.table::fread(file=paste(dir,export_file,dating_name,sep=""),sep=",")
   # hist(dates_pred$pred[dates_pred$pred<10])

#  import taxonomy assigned to sequences    names(dat)
dat<-data.table::fread(file=paste(dir,export_file,dat_name, sep=""), sep=",")
    
##   get read summary data --!!!! SV dat not correct- need fix  !!!!!
run_sum<-data.table::fread(file = paste(dir,export_file,"dammam_filter_summary_per_miseq.csv",sep=""), sep=",")

# get phylopic data - picture id number
pic<-openxlsx::read.xlsx(xlsxFile=pic_file) 
    
pic<- pic %>% 
  filter(Dammam == "yes") # !!!!!!

```

## basic_tidy

```{r tiding}
  hab_catc<-c("seagrass 1","seagrass 2", "mangrove 1", "mangrove 2","seagrass 3","seagrass 4","mangrove 3", "mangrove 4", "mangrove 5","seagrass 5","seagrass 6")
hab_cat_lev<-c("seagrass 1","seagrass 2","seagrass 3","seagrass 4","seagrass 5","seagrass 6", "mangrove 1", "mangrove 2","mangrove 3", "mangrove 4", "mangrove 5")

tl<-c("Superkingdom","Kingdom","Phylum", "Class", "Order", "Family", "Genus", "Species")

  loc1<- locations %>% # names(loc1)
    rename_all(tolower) %>% 
    arrange(-lat) %>% 
    mutate(core_id_pub=1:length(row.names(locations))) %>% 
    mutate(hab_cat=factor(hab_catc,levels=hab_cat_lev)) %>% 
    select(core,habitat, lat,lon,area,direction.loc)
  
##  sample data
sample_age1<- sample_age %>% 
  rename_all(tolower) %>% 
  filter(complete.cases(dating_type)) %>% 
  left_join(loc1)   # names(loc1)
  #mutate(habitat=factor(habitat), core_id_pub=factor(core_id_pub),      dating_type=factor(dating_type,levels=c("pb","c14")))

## predicted dates
dates_pred2 <- dates_pred %>%  
  select(id_dash, core, pred)

# make simple sample names
sample_names_good <- dat %>%    #   names(dat)
  filter(!duplicated(sample_id_u))  %>% 
  select(sample_id_u) %>% 
  mutate(num=1:length(row.names(.)), sample_id_meta=paste("x_",num,sep="")) %>% 
  select(-num)

###      tidy sequence data
 # unique(sample_age1$dating_type)
dat2<- dat %>%   # names(dat)   names(dates_pred)
  #
  #filter(reads>0) %>% #    !!!!!!!!!! remove, yes not ??????  !!!!!!!!!!!!!!!!!!!!!!
  #
  left_join(dates_pred2) %>%   ## names(dates_pred)
  left_join(loc1) %>%   ## names(loc1)
  dplyr::rename(ybp = pred) %>% 
  mutate(estimated_year=2016-ybp) %>% 
  mutate(type_3cat=forcats::fct_recode(type, "blank"="extraction blank", "blank" = "pcr blank", "mock" = "positive control")) %>% 
  mutate(ybp_cat=cut(ybp, breaks=c(-Inf,10,50,100,200,500, Inf), 
                        labels=c("<10","10-50","50-100","100-200","200-500",">500"))) %>%  
  mutate(ybp_cat=as.character(ybp_cat), type_3cat=as.character(type_3cat)) %>% 
  mutate(ybp_cat_blanks=if_else(is.na(ybp_cat), type_3cat  , ybp_cat )) %>% # bring blanks into age cat
    # unique(dat2$ybp_cat_blanks)
  mutate(ybp_cat_blanks=factor(ybp_cat_blanks, levels=c("<10","10-50","50-100","100-200","200-500",">500", "sample", "blank","mock"))) %>%
  mutate(direction.loc=factor(direction.loc, levels=c("North","Mid-north","Mid-south","South"))) %>% 
  mutate(percent_reads=reads/sample_sum_after_filt*100)  %>% 
  mutate(log_reads=log(reads+1), log_rare_reads=log(rare_reads+1), log_percent_reads=log(percent_reads+1) )  %>% 
  ## remove samples with minimal reads   !!!!!!!!!!!!!!!!!!!!!!!!!
  filter(sample_sum_after_filt>min_reads_per_miseq) %>% 
  ## get new/good sample names
  left_join(sample_names_good) %>% 
    mutate(primer=forcats::fct_recode(primer, "18S_V7"="euka02", "18S_V9"="18smini", "18S_V4"="18s_stoeck","CO1"="co1", "Vert"="vert", "RBCL"="rbclmini")) %>% 
  mutate(primer=factor(primer,levels=c("18S_V9","18S_V7","18S_V4","CO1","Vert","RBCL"))) %>% 
    tidyr::separate(lineage, into=tl, sep=";" , remove=FALSE) %>% 
  mutate(Superkingdom="Eukaryota") %>%   # fix in future, insect kingdom
  
  ## final filter, can't have NA's for comparisons
    filter(type=="sample") %>% 
    filter(complete.cases(habitat)) %>% 
  
## limit to specific miseq /  minboot data      unique(dat$primer_taxassign_id)
  filter(primer_taxassign_id== "co1;70" | primer_taxassign_id== "18s_stoeck;70" | primer_taxassign_id=="18smini;70" | primer_taxassign_id=="euka02;70" | primer_taxassign_id=="vert;70"| primer_taxassign_id=="rbclmini;70" ) %>% 
  # reset factors
  droplevels()    # levels(dat2$primer)  co1;insect
            

```


## remove edaphic
run if edaphic = yes
```{r edaphic}

if (edaphic== "yes"){
  
dat2<-dat2 %>% 
  filter(!grepl(paste(kingdom_remove, collapse="|"), Kingdom)) %>% 
  filter(!grepl(paste(phylum_remove, collapse="|"), Phylum)) %>% 
  filter(!grepl(paste(order_remove, collapse="|"), Order) )
}

  ## old from Dammam remove if moved on 
#filter(Phylum!="Annelida",Phylum!="Nematoda",Phylum!="Platyhelminthes",Phylum!="Rotifera",Phylum!="Nemertea" ) %>% 
#  filter(Kingdom!="Fungi") %>% 
#filter(Order!="Lucinoida",Order!="Nuculoida",Order!="Lamiales",Order!="Alismatales")
# infauna bivalve -Order== "Lucinoida"" , "Nuculoida" , seagrass and mangrove

```




## euk_phylum
tidy eukaryotes- sum by phylum, get 7 most abundant phyla
Sratipolot and dendogram infro from
 using https://paleolimbot.github.io/r4paleolim/strat-diagrams.html
 and http://paleoecologie.umontreal.ca/category/code/

```{r ei_format}
#  pool to phylum  !!!!!   get only needed data - keys, phylums, reads
euk_phylum<- dat2 %>%  #  names(dat2)
  select(ybp,habitat, core, primer ,Phylum, reads, rare_reads, percent_reads ) %>% 
  #filter(reads > 0) %>%  # remove with 0
  group_by(ybp, habitat, core, primer ,Phylum) %>% 
  summarise(reads_new=sum(reads), rare_reads_new=sum(rare_reads), percent_reads= sum(percent_reads), rich_reads= length(reads[reads>0]), rich_rare_reads= length(reads[rare_reads>0]), n=n() ) %>% ## cannot use reads again or will mess up calcs
  filter(Phylum !="NA") %>% 
  ungroup() %>% 
  ## drop primer ?
  filter(primer != "Vert" & primer != "RBCL") %>% 
  rename(reads=reads_new, rare_reads = rare_reads_new) %>% 
  droplevels()

# limit to 18S , get %'s and filter by number of sample reads     levels(dat_euk$primer)
euk_phylum2<-euk_phylum %>%   # names(dat3)  unique(dat3$habitat)
  filter(grepl("18S",primer)) %>% # keep only 18s primers
  mutate(primer=factor(primer)) %>% 
  group_by(primer, core, ybp) %>% 
  mutate(sum_euk_reads = sum(reads)) %>% 
  ungroup() %>% 
  mutate(percent_reads = (reads / sum_euk_reads)  * 100) %>% 
  #  !!!!!!  remove if sample ahs less than 100 euk reads  !!!
  filter(sum_euk_reads>=100)

## rank phylum by %
euk_phylum_rank <- euk_phylum2 %>%
  group_by(Phylum) %>%
  summarise(max_reads = max(percent_reads),mean = mean(percent_reads), n = n()) %>%
  arrange(desc(mean)) %>%
  filter( n > 20) %>% # filter phyla to only include those with this many samples
  pull(Phylum)

##  determine how many phyla to include
n<-7
euk_phylum_rank<-euk_phylum_rank[1:n]

## order by previous
euk_phylum_top7 <- euk_phylum2 %>%
  filter(Phylum %in% euk_phylum_rank) %>%
  mutate(Phylum = factor(Phylum, levels = euk_phylum_rank)) %>%
  arrange(Phylum) %>% 
   ungroup() 
  #mutate(ybp=as.integer(ybp))  # made some slices taht same !


```

## euk_order
tidy eukaryotes- sum by family, get 7 most abundant genus

```{r euk_order}
#  pool to genus  ,   get only needed data - keys, phylums, reads
euk_order<- dat2 %>%  #  names(dat2)
  select(ybp,habitat, core, primer, Phylum, Class, Order, reads, rare_reads, percent_reads ) %>% 
  group_by(ybp, habitat, core, primer ,Phylum, Class, Order) %>% 
  summarize(reads=sum(reads), rare_reads=sum(rare_reads), percent_reads= sum(percent_reads)) %>% 
  filter(Order != "NA") %>% 
  ungroup() %>% 
  ## drop primer ?
  filter(primer!="Vert" & primer!="RBCL") %>%  ##   !!!!!  limit  primers   !!!!!
  droplevels()

# simplify, get %'s and filter by number of sample reads     levels(dat_euk$primer)
euk_order2<-euk_order %>%   # names(dat3)  unique(dat3$habitat)
  unite("lineage", Phylum:Order, sep = ";", remove = FALSE) %>% 
  filter(grepl("18S",primer)) %>% #   !!!!!     keep only 18s primers     !!!!!!!
  mutate(primer=factor(primer)) %>% 
  group_by(primer, core, ybp) %>% # get sum per primer
  mutate(sum_euk_reads = sum(reads)) %>% 
  ungroup() %>% 
  mutate(percent_reads = (reads / sum_euk_reads)  * 100) %>% 
  #  !!!!!!  remove if sample ahs less than 100 euk reads  !!!
  filter(sum_euk_reads>=100)

## rank Genus by %
euk_order_rank <- euk_order2 %>%
  group_by(lineage) %>%
  summarise(max_reads = max(percent_reads),mean=mean(percent_reads), median=median(percent_reads), n=n()) %>%
  arrange(desc(mean)) %>%   # !!  either mean or median, similar but depnds on skew
  filter( n>5) %>% # filter phyla to only include those with this many samples
  pull(lineage)

##  determine how many phyla to include
n<-7
euk_order_rank<-euk_order_rank[1:n]
euk_order_rank_cl<-gsub("^.*\\;","", euk_order_rank)

## order by previous
euk_order_top7 <- euk_order2 %>%
   filter(lineage %in% euk_order_rank) %>%
  mutate(Order=gsub("^.*\\;","", lineage)) %>% 
  mutate(Order = factor(Order, levels = euk_order_rank_cl)) %>%
  arrange(lineage) %>% 
   ungroup() # %>% 


```


## euk_genus
tidy eukaryotes- sum by family, get 7 most abundant genus

```{r ei_format}
#  pool to genus  ,   get only needed data - keys, phylums, reads
euk_genus<- dat2 %>%  #  names(dat2)
  select(ybp,habitat, core, primer, Phylum, Class, Order, Family, Genus, reads, rare_reads, percent_reads ) %>% 
  group_by(ybp, habitat, core, primer ,Phylum, Class, Order, Family, Genus) %>% 
  summarize(reads=sum(reads), rare_reads=sum(rare_reads), percent_reads= sum(percent_reads)) %>% 
  filter(Genus != "NA") %>% 
  ungroup() %>% 
  ## drop primer ?
  filter(primer!="Vert" & primer!="RBCL") %>%  ##   !!!!!  limit  primers   !!!!!
  droplevels()

# simplify, get %'s and filter by number of sample reads     levels(dat_euk$primer)
euk_genus2<-euk_genus %>%   # names(dat3)  unique(dat3$habitat)
  unite("lineage", Phylum:Genus, sep = ";", remove = FALSE) %>% 
  filter(grepl("18S",primer)) %>% #   !!!!!     keep only 18s primers     !!!!!!!
  mutate(primer=factor(primer)) %>% 
  group_by(primer, core, ybp) %>% 
  mutate(sum_euk_reads = sum(reads)) %>% 
  ungroup() %>% 
  mutate(percent_reads = (reads / sum_euk_reads)  * 100) %>% 
  #  !!!!!!  remove if sample ahs less than 100 euk reads  !!!
  filter(sum_euk_reads>=100)

## rank Genus by %
euk_genus_rank <- euk_genus2 %>%
  group_by(lineage) %>%
  summarise(max_reads = max(percent_reads),mean=mean(percent_reads), median=median(percent_reads), n=n()) %>%
  arrange(desc(mean)) %>%   # !!  either mean or median, similar but depnds on skew
  filter( n>5) %>% # filter phyla to only include those with this many samples
  pull(lineage)

##  determine how many phyla to include
n<-7
euk_genus_rank<-euk_genus_rank[1:n]
euk_genus_rank_cl<-gsub("^.*\\;","", euk_genus_rank)

## order by previous
euk_genus_top7 <- euk_genus2 %>%
   filter(lineage %in% euk_genus_rank) %>%
  mutate(Genus=gsub("^.*\\;","", lineage)) %>% 
  mutate(Genus = factor(Genus, levels = euk_genus_rank_cl)) %>%
  arrange(lineage) %>% 
   ungroup() # %>% 


```

## metaz_phylum
limit to metazoans- sum by phylum, get 7 most abundant phyla

```{r metaz_format}
# limit to metaz and recalc percent, get richness and diversity

## make pattern for filtering
metaz_pat<-c("Annelida","Arthropoda","Chordata","Cnidaria","Ctenophora","Echinodermata","Gastrotricha","Mollusca","Nematoda","Orthonectida","Platyhelminthes","Porifera","Rotifera","Sipuncula","Tardigrada","Urochordata","Chaetognatha","Craniata","Nemertea","Bryozoa","Hemichordata","Brachiopoda","Entoprocta","Cephalochordata","Entoprocta","Gnathostomulida", "Xenacoelomorpha") 
metaz_pat<-paste(metaz_pat, collapse = "|")

metaz_phylum <-euk_phylum %>%   # names(euk_phylum)  unique(dat3$habitat) 
  filter(grepl(metaz_pat, Phylum))  %>% 
  group_by(primer, core, ybp) %>% 
  mutate(sum_metaz_reads = sum(reads)) %>% 
  ungroup() %>% 
  mutate(percent_reads = (reads / sum_metaz_reads)  * 100) %>% 
  #  !!!!!!  remove if sample ahs less than 100 metazoan reads  !!!
  filter(sum_metaz_reads>=100)

## rank phylum by %
metaz_phylum_rank <- metaz_phylum %>%
  group_by(Phylum) %>%
  summarise(max_reads = max(percent_reads), mean=mean(percent_reads), n=n()) %>%
  arrange(desc(mean)) %>%
  filter( n>20) %>% # filter phyla to only include those with this many samples
  pull(Phylum)

##  determine how many phyla to include
n<-7
metaz_phylum_rank<-metaz_phylum_rank[1:n]

## limit and order by previous
metaz_phylum_top7 <- metaz_phylum %>%
  filter(Phylum %in% metaz_phylum_rank) %>%
  mutate(Phylum = factor(Phylum, levels = metaz_phylum_rank)) %>%
  arrange(Phylum) %>% 
   ungroup() # %>% 
  #mutate(ybp=as.integer(ybp))  # made some slices taht same !

## get rads per sample for summary plot     names(dat_metaz)      names(dat_metaz_reads_sam)

dat_metaz_reads_sam <- metaz_phylum %>% 
  mutate(id=paste(ybp,habitat,core,primer, sep="-")) %>% 
  filter(!duplicated(id)) %>% 
  select(-reads, -rare_reads, -percent_reads,-Phylum, -id) %>% 
  dplyr::rename(resp = sum_metaz_reads) %>% 
  mutate(fac="Metaz_reads_sample") 
```

## metaz_order
limit to metazoans- sum by genus, get 7 most abundant genus

```{r metaz_order}
# limit to metaz and recalc percent, get richness and diversity

# limit to metazoan phyla then get reads per sample and % per taxa
metaz_order<-euk_order %>%   # names(euk_genus)  unique(dat3$habitat)
  filter(grepl(metaz_pat, Phylum))  %>%  # limit to metaz phyla
  unite("lineage", Phylum:Order, sep = ";", remove = FALSE) %>% 
  group_by(primer, core, ybp) %>% 
  mutate(sum_metaz_reads = sum(reads)) %>% 
  ungroup() %>% 
  mutate(percent_reads = (reads / sum_metaz_reads)  * 100) %>% 
  #  !!!!!!  remove if sample ahs less than 100 metazoan reads  !!!
  filter(sum_metaz_reads >= 100)

## rank order by %
metaz_order_rank <- metaz_order %>%
  group_by(lineage, Phylum, Class, Order) %>%
  summarise(max_reads = max(percent_reads),mean=mean(percent_reads), n=n()) %>%
  arrange(desc(mean)) %>%
  filter( n>5) %>% # filter phyla to only include those with this many samples
  pull(lineage)

##  determine how many phyla to include
n<-7
metaz_order_rank<-metaz_order_rank[1:n]
metaz_order_rank_cl<-gsub("^.*\\;","", metaz_order_rank)

## limit to top 7 and arrange
metaz_order_top7 <- metaz_order %>%
  filter(lineage %in% metaz_order_rank) %>%
  mutate(Order=gsub("^.*\\;","", lineage)) %>% 
  mutate(Order = factor(Order, levels = metaz_order_rank_cl)) %>%
  arrange(lineage) %>% 
   ungroup() # %>% 
  #mutate(ybp=as.integer(ybp))  # made some slices taht same !

```

## metaz_genus
limit to metazoans- sum by genus, get 7 most abundant genus

```{r metaz_format}
# limit to metaz and recalc percent, get richness and diversity


# limit to metazoan phyla then get reads per sample and % per taxa
metaz_genus<-euk_genus %>%   # names(euk_genus)  unique(dat3$habitat)
  filter(grepl(metaz_pat, Phylum))  %>%  # limit to metaz phyla
  unite("lineage", Phylum:Genus, sep = ";", remove = FALSE) %>% 
  group_by(primer, core, ybp) %>% 
  mutate(sum_metaz_reads = sum(reads)) %>% 
  ungroup() %>% 
  mutate(percent_reads = (reads / sum_metaz_reads)  * 100) %>% 
  #  !!!!!!  remove if sample ahs less than 100 metazoan reads  !!!
  filter(sum_metaz_reads>=100)

## rank Genus by %
metaz_genus_rank <- metaz_genus %>%
  group_by(lineage, Phylum, Class, Order, Family, Genus) %>%
  summarise(max_reads = max(percent_reads),mean=mean(percent_reads), n=n()) %>%
  arrange(desc(mean)) %>%
  filter( n>5) %>% # filter phyla to only include those with this many samples
  pull(lineage)

##  determine how many phyla to include
n<-7
metaz_genus_rank<-metaz_genus_rank[1:n]
metaz_genus_rank_cl<-gsub("^.*\\;","", metaz_genus_rank)

## limit to top 7 and arrange
metaz_genus_top7 <- metaz_genus %>%
  filter(lineage %in% metaz_genus_rank) %>%
  mutate(Genus=gsub("^.*\\;","", lineage)) %>% 
  mutate(Genus = factor(Genus, levels = metaz_genus_rank_cl)) %>%
  arrange(lineage) %>% 
   ungroup() # %>% 
  #mutate(ybp=as.integer(ybp))  # made some slices taht same !

```


## habitat
get seagrass and mangrove sequences
```{r habitat}
dat_hab<- dat2 %>%  #  names(dat2)
  
    filter(primer_taxassign_id== "18s_stoeck;70" | primer_taxassign_id=="18smini;70" | primer_taxassign_id=="euka02;70" |primer_taxassign_id=="rbclmini;70" )  %>% 
  
  select(ybp,habitat, core, primer ,Order, reads, rare_reads, percent_reads ) %>% 
  group_by(ybp, habitat, core, primer ,Order) %>% 
  summarize(reads=sum(reads), rare_reads=sum(rare_reads), percent_reads= sum(percent_reads)) %>% 
  filter(Order!="NA") %>% 
  ungroup() %>% 
  ## drop primer ?
  filter(primer!="Vert" & primer!="CO1") %>% 
  droplevels() # sort(unique(dat_hab$Order))

# limit to metaz and recalc percent, get richness and diversity
## make pattern for filtering
hab_pat<-c("Lamiales", "Alismatales") 
hab_pat<-paste(hab_pat, collapse = "|")

dat_hab2<-dat_hab %>%   # names(dat3)
  filter(grepl(hab_pat, Order))  %>% 
  mutate(fac=Order, resp=percent_reads)  %>% 
  #mutate(fac=paste(habitat,Order,sep="_"))  %>% 
  
  select(-percent_reads,-Order,-reads,-rare_reads)
  
```



## diversity
calculate diverstiy for only metazoans
```{r diverstiy}

# fac includes -per sample     reads, mean hab reads ,rich, div

##    get diversity of metazoans
div<- dat2 %>%   # names(dat2)
  ungroup() %>% 
  filter(primer_taxassign_id== "co1;70" | primer_taxassign_id== "18s_stoeck;70" | primer_taxassign_id=="18smini;70" | primer_taxassign_id=="euka02;70"  ) %>% 
  
  filter(grepl(metaz_pat, Phylum))  %>% # only metazoans
  group_by(primer, core, ybp) %>% 
  mutate(sum_metaz_reads = sum(reads)) %>% # get new % per miseq
  ungroup() %>% 
  mutate(per_metaz_reads=(reads/sum_metaz_reads)*100) %>% 
  select(ybp, habitat, core, primer ,Phylum:Species, per_metaz_reads, rare_reads) %>% 
  group_by(ybp, habitat, primer, core) %>% 
  summarize( 
    metaz_richness = length(rare_reads[rare_reads>0]),   
    metaz_shannon = sum(  (-1*(log( (per_metaz_reads/100) + 0.01 ))) * (per_metaz_reads/100+ 0.01) ),
    metaz_simpsons = 1 - sum((per_metaz_reads / 100) ^ 2),
    metaz_hill_n2 = 1 / sum((per_metaz_reads / 100) ^ 2))  %>% 
    ungroup() %>% 
    mutate(ybp=as.integer(ybp))

## make fac by gathering
metaz_div <- div %>% 
  select(-metaz_simpsons, -metaz_hill_n2)  %>% 
  gather(key="fac", value="resp", "metaz_richness", "metaz_shannon")

```


## tidy_sum_strat
```{r tidy_sum_strat}

lev_sum<-c("sample_sum_after_filt","Metaz_reads_sample", "rich_rare","shan_div", "metaz_richness","metaz_shannon")

sum1<-dat2 %>% ## names(dat2)   names(dat_metaz_reads_sam)  names(sum1)     names(div2)
    filter(primer_taxassign_id== "co1;70" | primer_taxassign_id== "18s_stoeck;70" | primer_taxassign_id=="18smini;70" | primer_taxassign_id=="euka02;70") %>% 
  group_by(sample_id_u)  %>% 
  mutate(sample_sum_after_fil_rare=sum(rare_reads)) %>% 
  ungroup() %>% 
    select(ybp,habitat,core,primer,sample_sum_after_filt,shan_div,rich_rare) %>% 
    mutate(id=paste(ybp,habitat,core,primer, sep="-")) %>% 
    filter(!duplicated(id)) %>% 
    select(-id) %>% 
    gather(key="fac", value="resp", "sample_sum_after_filt", "shan_div", "rich_rare") %>% 
  # get other tables
  bind_rows(metaz_div) %>% ## metaz_div
   bind_rows(dat_metaz_reads_sam) %>%   # metaz reads
    #bind_rows(dat_hab2) %>%   # habitat reads
  
  # make fac a factor and organize
  mutate(fac=factor(fac, levels=lev_sum)) %>%    # unique(sum1$fac)
  mutate(fac=forcats::fct_recode(fac, "Reads"="sample_sum_after_filt" , "Metaz reads"="Metaz_reads_sample", "18S richness" = "rich_rare", "Metaz richness" = "metaz_richness" , "18S diversity" = "shan_div", "Metaz diversity" = "metaz_shannon")) %>% 
  #  remove co1 from Richness  !!
  filter(!(fac =="18S richness" & primer =="CO1") )  %>%  # unique(sum1$primer)
  filter(!(fac =="18S diversity" & primer =="CO1") ) 

```


## euk_dendo
 from http://paleoecologie.umontreal.ca/category/code/
```{r euk_dendo}
## dendogram for 18s primers
# dat_metaz  names(dat_metaz)

##########  for euk by phylum
euk_phylum_dist<- euk_phylum2 %>% 
  select(ybp,habitat,core,primer,Phylum,percent_reads) %>% 
  filter(ybp<max_age_depth) %>% 
  spread(key=Phylum, value=percent_reads, fill=0) 

euk_phylum_ddata<-dendo_prep(x=euk_phylum_dist)

##########  for euk by order
euk_order_dist<- euk_order2 %>% 
  select(ybp,habitat,core,primer,lineage,percent_reads) %>% 
  filter(ybp<max_age_depth) %>% 
  spread(key=lineage, value=percent_reads, fill=0) 

euk_order_ddata<-dendo_prep(x=euk_order_dist)

##########  for euk by genus
euk_genus_dist<- euk_genus2 %>% 
  select(ybp,habitat,core,primer,lineage,percent_reads) %>% 
  filter(ybp<max_age_depth) %>% 
  spread(key=lineage, value=percent_reads, fill=0) 

euk_genus_ddata<-dendo_prep(x=euk_genus_dist)

```



## plot_euk
make strati plot of 7 greatest phyla and cluster of all phyla, then same for genus
```{r plot_euk}
#  plot plylums     names(euk_phylum_top7)
pp1<-strati_plot_7(x=euk_phylum_top7, y=euk_phylum_ddata, taxa_group=euk_phylum_top7$Phylum,
                   pic_rank=euk_phylum_rank, cluster_adj=0.0, add_pic="yes")
  ggsave(plot=pp1, file=paste(dir,plot_file,"/strati_euk_phylum_",max_age_depth_char,"deep.pdf",sep=""), width = 26, height = 10, units = "cm")
  
#  plot order
pp2<-strati_plot_7(x=euk_order_top7, y=euk_order_ddata, taxa_group=euk_order_top7$Order,
                   pic_rank=euk_order_rank_cl, cluster_adj=0.0, add_pic="yes")
  ggsave(plot=pp2, file=paste(dir,plot_file,"/strati_euk_order_",max_age_depth_char,"deep.pdf",sep=""), width = 26, height = 10, units = "cm")

#  plot genus
pp3<-strati_plot_7(x=euk_genus_top7, y=euk_genus_ddata, taxa_group=euk_genus_top7$Genus,
                   cluster_adj=0.0, add_pic="no")
  ggsave(plot=pp3, file=paste(dir,plot_file,"/strati_euk_genus_",max_age_depth_char,"deep.pdf",sep=""), width = 26, height = 10, units = "cm")
  

```


## plot_euk_rich
make strati plot of 7 greatest phyla richness to see what phyla causing increase
```{r plot_euk_rich}
  # plot stratiplot of top 7 
x<-euk_phylum_top7
max_age = max_age_depth
abund_lim = c(0, 22)
abund_breaks = c(1,10,20)
x$taxa_group1<-x$Phylum

richr<-ggplot(x, aes(x = ybp, y = rich_rare_reads)) +
   geom_point(aes(x = ybp, y = rich_rare_reads, colour=habitat, shape=primer), size = 2) +
   geom_smooth(aes(x = ybp, y = rich_rare_reads), span = 0.9, colour="grey", se=TRUE)   +
   geom_smooth(aes(x = ybp, y = rich_rare_reads, colour=habitat), span = 0.9, se=FALSE)   + #loess  se=FALSE
   ## if less than 1000 observations uses loess
  # facet by taxon, keeping distance on the x axis comparable
  facet_grid(~taxa_group1, scales = "fixed", space = "fixed") +
  # have the same breaks for all x axes
  scale_y_continuous(limits = abund_lim,  expand = c(0,0), breaks=abund_breaks , minor_breaks=NULL) +
  scale_x_reverse(limits = c(max_age, 0)) +
  scale_color_manual(values=c("tan3","seagreen3")) +
  #scale_colour_brewer(palette = "Set1") +
  scale_shape(solid=FALSE) +
    labs(y = "Richness", x = "Years before present")  +
  coord_flip() +
   theme(legend.position="bottom") +
    theme_bw() + # classic -no grids or bw
  theme(panel.spacing = unit(0.7, "lines")) + # increase margins
  annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf) +   # x on left - annotate("segment", x=-Inf, xend=-Inf, y=-Inf, yend=Inf)
  annotate("segment", x=Inf, xend=Inf, y=-Inf, yend=Inf) 

# ggsave(plot=richr, file=paste(dir,plot_file,"/strati/strati_euk_phyla_richness.pdf",sep=""), width = 20, height = 10, units = "cm")

#### mess see changes in Bacillariophyta
bacill<- dat2 %>%  #  names(dat2)
  #select(ybp,habitat, core, primer ,Phylum, reads, rare_reads, percent_reads ) %>% 
  filter(Phylum == "Bacillariophyta") %>% 
  filter(Phylum != "NA") %>% 
  ungroup() %>% 
  ## drop primer ?
  filter(primer != "Vert" & primer != "RBCL") %>% 
  filter(grepl("18S",primer)) %>% # keep only 18s primers
  filter(rare_reads>0, ybp>40)%>%  #, ybp<1
  droplevels() 

 unique(bacill$Species)
 
 mes<-bacill %>% 
    filter(!duplicated(sample_id_u))  # hist(mes$ybp)  median(mes$ybp)
 
# 38 species total 36<1  36<10  30>100   22>200   37<40, 31>40-median



```


## metaz_dendo
 from http://paleoecologie.umontreal.ca/category/code/
```{r meta_dendo}
## dendogram for metazoan
# dat_metaz  names(dat_metaz)
metaz_phylum_dist<- euk_phylum2 %>% 
  select(ybp,habitat,core,primer,Phylum,percent_reads) %>% 
  filter(ybp < max_age_depth) %>% 
  spread(key=Phylum, value=percent_reads, fill=0) 

metaz_phylum_ddata<-dendo_prep(x=metaz_phylum_dist)

#######   for order
# dat_metaz  names(metaz_order2)
metaz_order_dist<- metaz_order %>% 
  select(ybp,habitat,core,primer,lineage,percent_reads) %>% 
  filter(ybp < max_age_depth) %>% 
  spread(key=lineage, value=percent_reads, fill=0) 

metaz_order_ddata<-dendo_prep(x=metaz_order_dist)

#######   for genus
# dat_metaz  names(metaz_genus2)
metaz_genus_dist<- metaz_genus %>% 
  select(ybp,habitat,core,primer,lineage,percent_reads) %>% 
  filter(ybp < max_age_depth) %>% 
  spread(key=lineage, value=percent_reads, fill=0) 

metaz_genus_ddata<-dendo_prep(x=metaz_genus_dist)

```


## plot_metaz

```{r plot_metaz}

#  plot plylums
ff<-strati_plot_7(x=metaz_phylum_top7, y=metaz_phylum_ddata, taxa_group=metaz_phylum_top7$Phylum,
                   pic_rank=metaz_phylum_rank, cluster_adj=0.0, add_pic="yes")
ggsave(plot=ff, file=paste(dir,plot_file,"/strati_metaz_phylum_",max_age_depth_char,"deep.pdf",sep=""), width = 26, height = 10, units = "cm")

#  plot order
ff<-strati_plot_7(x=metaz_order_top7, y=metaz_order_ddata, taxa_group=metaz_order_top7$Order,
                   pic_rank=metaz_order_rank_cl, cluster_adj=0.0, add_pic="yes")
ggsave(plot=ff, file=paste(dir,plot_file,"/strati_metaz_order_",max_age_depth_char,"deep.pdf",sep=""), width = 26, height = 10, units = "cm")

#  plot genus
ff<-strati_plot_7(x=metaz_genus_top7, y=metaz_genus_ddata, taxa_group=metaz_genus_top7$Genus,
                   pic_rank=metaz_order_rank_cl, cluster_adj=0.0, add_pic="no")
ggsave(plot=ff, file=paste(dir,plot_file,"/strati_metaz_genus_",max_age_depth_char,"deep.pdf",sep=""), width = 26, height = 10, units = "cm")

```



## plot_summar
```{r plot_summary}  

## plot reads
datp<-sum1 %>%   # unique(sum1$fac)
  #filter(fac=="Reads" | fac=="Metaz reads") 
  mutate(resp=if_else(fac=="Reads", log(resp+1), resp)) %>% 
  mutate(resp=if_else(fac=="Metaz reads", log(resp+1), resp)) 
datp$dep<-datp$ybp
bb<-c(0,1,5,10)

age_axis_trans<-"yes" ; x_breaks = c(1,25,100,250,500,750)

read_sum<-ggplot(datp, aes(x = dep, y = resp)) +
  geom_point(aes(x = dep, y = resp, colour=habitat, shape=primer), size = 2) +
  geom_smooth(aes(x = dep, y = resp), span = 0.9, colour="grey")   +
  geom_smooth(aes(x = dep, y = resp, colour=habitat), span = 0.9, se=FALSE)   + #loess  se=FALSE
  facet_grid(~fac, space = "fixed", scales="free") +  # , scales = "fixed"
  scale_y_continuous(minor_breaks=NULL,  expand = c(0,0)) + { #, limits = c(0, NA)
    ## if then for trans yes no
  if(age_axis_trans=="yes") scale_x_continuous(trans=reverse_sqrt_trans(), limits = c(max_age, 0), breaks = x_breaks, expand = c(0.03,0), minor_breaks=NULL) else scale_x_reverse(limits = c(max_age_dat, 0) ) 
    } +
  
  scale_color_manual(values=c("tan3","seagreen3")) +
  scale_shape(solid=FALSE) +
  labs(y = "", x = "Years before present")  +
  coord_flip( ) +
     #theme(legend.position="right") +
    theme_bw() + # classic -no grids or bw
  # increase margins and reduce font of plot labels wihtin facet_grid
  theme(panel.spacing = unit(0.5, "lines"), strip.text.x = element_text(size = 8)) + 
  annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf) +  
  annotate("segment", x=Inf, xend=Inf, y=-Inf, yend=Inf)

##   plot summary
ggsave(read_sum, file=paste(dir,plot_file,"/strati_summary_",max_age_depth_char,"deep.pdf",sep=""), width = 20, height = 10, units = "cm")
#ggsave(file=paste(dir,plot_file,"/strati_summary_",max_age_depth_char,"deep.tiff",sep=""), width = 20, height = 10, units = "cm", dpi="print")
```




## export_data
```{r save_data}

#  data.table::fwrite(edna_slice,paste(dir,export_file,export_name,sep=""),row.names=F, sep=",")

```

